### PR TITLE
CP-49367  Implement xenserver_pif provider data source

### PR DESCRIPTION
**Pull request by acefei**
```
source .env && TF_ACC=1 go test ./xenserver/ -v   -timeout 120m
=== RUN   TestAccNetworkResource
--- PASS: TestAccNetworkResource (50.38s)
=== RUN   TestAccPifDataSource
--- PASS: TestAccPifDataSource (19.44s)
=== RUN   TestAccSRDataSource
--- PASS: TestAccSRDataSource (18.06s)
=== RUN   TestAccVMResource
--- PASS: TestAccVMResource (52.08s)
PASS
ok      terraform-provider-xenserver/xenserver  139.970s
```

```
pif_data_out = tolist([
  {
    "bond_master_of" = tolist([])
    "bond_slave_of" = "OpaqueRef:NULL"
    "capabilities" = tolist([
      "sriov",
    ])
    "currently_attached" = true
    "device" = "eth0"
    "disallow_unplug" = false
    "dns" = "10.71.56.11"
    "gateway" = "10.71.56.1"
    "host" = "OpaqueRef:041ade78-40d3-1ecd-3d3c-b7b7e91e4f76"
    "igmp_snooping_status" = "disabled"
    "ip" = "10.71.56.50"
    "ip_configuration_mode" = "DHCP"
    "ipv6" = tolist([])
    "ipv6_configuration_mode" = "None"
    "ipv6_gateway" = ""
    "mac" = "0c:c4:7a:e7:b5:e8"
    "managed" = true
    "management" = true
    "mtu" = 1500
    "netmask" = "255.255.248.0"
    "network" = "OpaqueRef:ce271ed2-87c5-ee24-0e71-0fead2417986"
    "other_config" = tomap({})
    "pci" = "OpaqueRef:a7e83bf8-0967-e886-818a-e259bda0d98e"
    "physical" = true
    "primary_address_type" = "IPv4"
    "properties" = tomap({
      "gro" = "on"
    })
    "sriov_logical_pif_of" = tolist([])
    "sriov_physical_pif_of" = tolist([])
    "tunnel_access_pif_of" = tolist([])
    "tunnel_transport_pif_of" = tolist([])
    "uuid" = "d9e981a6-7251-8794-7bbd-acb8662b9633"
    "vlan" = -1
    "vlan_master_of" = "OpaqueRef:NULL"
    "vlan_slave_of" = tolist([])
  },
])
vm_out = {
  "id" = "c43e9f42-c48d-2dce-19cd-2a2b7ba447ef"
  "name_label" = "Test CentOS VM"
  "other_config" = tomap({
    "flag" = "1"
  })
  "snapshots" = tolist([])
  "template_name" = "CentOS 7"
}
```

### CODE REVIEW

File: xenserver/pif_data_source.go [hunk 1]
```diff
@@ -51,9 +198,16 @@ func (d *pifDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, re
 				Optional:            true,
 			},
 			"network": schema.StringAttribute{
```
Comment (xueqingz) [inline xenserver/pif_data_source.go:200]: why not just use "uuid" here?

File: xenserver/pif_data_source.go [hunk 2]
```diff
@@ -35,12 +34,160 @@ func (d *pifDataSource) Metadata(_ context.Context, req datasource.MetadataReque
...
+	}
+}
+
 func (d *pifDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
 	resp.Schema = schema.Schema{
-		// This description is used by the documentation generator and the language server.
-		MarkdownDescription: "PIF data source",
-
+		MarkdownDescription: "Provides information about a physical network interface (PIF) of XenServer",
 		Attributes: map[string]schema.Attribute{
```
Comment (xueqingz) [inline xenserver/pif_data_source.go:191]: there is another key "management = true" in main.tf, but I didn't find in schema.

File: xenserver/pif_data_source.go [hunk 3]
```diff
@@ -35,12 +34,160 @@ func (d *pifDataSource) Metadata(_ context.Context, req datasource.MetadataReque
 	resp.TypeName = req.ProviderTypeName + "_pif"
 }
 
-// Schema defines the schema for the data source.
+func PifSchema() map[string]schema.Attribute {
```
Comment (xueqingz) [inline xenserver/pif_data_source.go:37]: PifSchema() -> pifSchema()

File: xenserver/pif_data_source.go [hunk 4]
```diff
@@ -92,14 +246,33 @@ func (d *pifDataSource) Read(ctx context.Context, req datasource.ReadRequest, re
 		return
```
Comment (xueqingz) [inline xenserver/pif_data_source.go:246]: Unable to read PIF records

**Comment by xueqingz** [comment]
Could you also update the pif_data_source_test.go  base on your changes, include the new fields.

File: xenserver/pif_data_source.go [hunk 5]
```diff
@@ -51,9 +198,16 @@ func (d *pifDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, re
 				Optional:            true,
 			},
 			"network": schema.StringAttribute{
-				MarkdownDescription: "UUID of the virtual network to which this PIF is connected",
+				MarkdownDescription: "The UUID of the virtual network to which this PIF is connected",
 				Computed:            true,
```
Comment (xueqingz) [inline xenserver/pif_data_source.go:202]: Why this is computed value?

File: xenserver/pif_data_source.go [hunk 6]
```diff
@@ -35,12 +34,160 @@ func (d *pifDataSource) Metadata(_ context.Context, req datasource.MetadataReque
...
+			Computed:            true,
+			ElementType:         types.StringType,
+		},
+		"other_config": schema.MapAttribute{
+			MarkdownDescription: "Additional configuration",
+			Computed:            true,
+			ElementType:         types.StringType,
+		},
+		"disallow_unplug": schema.BoolAttribute{
+			MarkdownDescription: "Prevent this PIF from being unplugged; set this to notify the management tool-stack that the PIF has a special use and should not be unplugged under any circumstances (e.g. because you&apos;re running storage traffic over it)",
```
Comment (xueqingz) [inline xenserver/pif_data_source.go:123]: format error here, `because you&apos;re `

Comment (acefei) [reply xenserver/pif_data_source.go:123]: This copied from go SDK, so you might update there either.

File: xenserver/pif_data_source.go (hunk 2 above)
Comment (acefei) [reply xenserver/pif_data_source.go:191]: in line 47

File: xenserver/pif_data_source.go [hunk 7]
```diff
@@ -86,20 +240,39 @@ func (d *pifDataSource) Read(ctx context.Context, req datasource.ReadRequest, re
...
+
+		if !data.Management.IsNull() && pifRecord.Management != data.Management.ValueBool() {
+			continue
+		}
+
+		var pifData pifRecordData
+		err = updatePIFRecordData(ctx, pifRecord, &pifData)
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Unable to update SR record data",
```
Comment (xueqingz) [inline xenserver/pif_data_source.go:267]: "Unable to update PIF record data"

**Review by xueqingz — APPROVED** [review]

File: docs/data-sources/pif.md [hunk 8]
```diff
@@ -30,7 +30,48 @@ output "pif_data_out" {
...
+### Nested Schema for `data_items`
+
+Read-Only:
+
+- `bond_master_of` (List of String) Indicates this PIF represents the results of a bond
+- `bond_slave_of` (String) Indicates which bond this interface is part of
+- `capabilities` (List of String) Additional capabilities on the interface.
+- `currently_attached` (Boolean) True if this interface is online
+- `device` (String) The machine-readable name of the physical interface (PIF) (e.g. eth0)
+- `disallow_unplug` (Boolean) Prevent this PIF from being unplugged; set this to notify the management tool-stack that the PIF has a special use and should not be unplugged under any circumstances (e.g. because you're running storage traffic over it)
```
Comment (lindig) [inline docs/data-sources/pif.md:49]: I would like to see MD documents not exceeding 80 characters in line length such that reviews, diffs, and reading them in a terminal is supported.

Comment (acefei) [reply docs/data-sources/pif.md:49]: This is a generated doc from Go, and it can't be broken up into multiple lines, but we can try to make it shorter.

**Pull request merged**